File: submissions/Python3/59.py
class Solution:
    def generateMatrix(self, n):
        """
        :type n: int
        :rtype: List[List[int]]
        """
        ret = [[0 for i in range(n)] for j in range(n)]
        top, down, left, right= 0, n, 0, n
        m = 1
        while left < right and top < down:
            for y in range(left, right):
                ret[top][y] = m
                m += 1
            top += 1
            for x in range(top, down):
                ret[x][right - 1]= m
                m += 1
            right -= 1
            for y in range(right - 1, left - 1, -1):
                ret[down - 1][y]= m
                m += 1
            down -= 1
            for x in range(down - 1, top - 1, -1):
                ret[x][left]= m
                m += 1
            left += 1
        return ret

File: LeetCode/cpp/59.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> ret(n, vector<int>(n, 0));
        int N = pow(n, 2), m = 1;
        int left = 0, right = n - 1, top = 0, down = n - 1;
        while (top <= down && left <= right) {
            for (int y = left; y <= right; ++y)
                ret[top][y] = m++;
            ++top;
            for (int x = top; x <= down; ++x)
                ret[x][right] = m++;
            --right;
            for (int y = right; y >= left; --y)
                ret[down][y] = m++;
            --down;
            for (int x = down; x >= top; --x)
                ret[x][left] = m++;
            ++left;
        }
        return ret;
    }
};
